directus/directus
# API / Cache: Improve schema memory resources

## Scope
When doing some tests with a huge schema (>6000 entities, collections + fields + relations) I have found out that my Directus instance was running out of memory.
<details><summary>
To better understand what was causing this issue I have used Memory inspector and took some snapshots.
</summary>
I am not able to include them here as it has private data
</details>

When I took a look into the memory snapshot, I saw the schema repeated multiple times but without some properties.
I have found out that the [setLocalSchemaCache](https://github.com/joselcvarela/directus/blob/api/cache/improve-schema-memory/api/src/cache.ts#L121-L125) was the culprit.
More specifically, the schema is being cached in memory using Keyv.
But Keyv does serialize / deserialize all data that passes through it.
Since there's no specification, the default serializer / deserializer are being used.
[As we can see on the serializer](https://github.com/jaredwray/keyv/blob/main/packages/serialize/src/index.ts#L5-L53) it is traversing the entire object and store chunks of it as strings in memory.
This causes the data being duplicated in memory.
But this is not even necessary, because the schema is already in memory, so we just need to store it's reference, no need to serialize it.

In this PR, we just change the serialize / deserialize to just store the reference of the schema object.
Since Keyv is using a [Map](https://github.com/jaredwray/keyv/blob/main/packages/keyv/src/index.ts#L106) there should be no issues.

What's changed:

- Improvement on memory consumption storing schema into memory

## Potential Risks / Drawbacks

- If Keyv decides to use anything else other than a `Map` or any other structure that can hold object, this may break

## Review Notes / Questions

- Testing this can be somehow difficult but I am leaving some steps below

## Testing
In order to test this fix, I have:
1. Run Postgres and Redis via Docker Compose
2. Run bootstrap
3. Use cache, cache system and auto purge cache
4. Use Directus 10.13.3 via Docker
5. Use command `node --inspect=0.0.0.0 cli.js start`. This allows to investigate Memory.
6. Use DevTools to track memory usage
7. Run K6 to fetch data from multiple collections
8. At same time, do multiple sort fields in Data Model
9. See memory increasing and container run out of memory
10. Build a new image from 7dc377f8fd (tag v10.13.3) but with the fix implemented
11. Use new build image and repeat process from 5 until 8
12. After doing multiple sort fields (more than previous) the container was still running as expected with low memory
___

Fixes #24035
Closes SER-440
